### PR TITLE
A subinterpreter-in-thread spawning backend

### PR DESCRIPTION
**Pull request by goodboy**
### Motivation

This PR is **Phase B.2** of the multi-backend spawn rework tracked
under #379: it lands the first member of the `subint` family — a
single-process spawn backend that runs each sub-actor inside a
CPython PEP 734 sub-interpreter driven by a dedicated OS thread,
while reusing tractor's existing IPC handshake (UDS/TCP) for the
parent↔child channel. The result is multiple `Actor`s in **one OS
process** with state isolation and faster startup than
`mp_spawn`/`mp_forkserver`, but **without** the per-interp-GIL
parallelism that's still gated upstream on `msgspec` PEP 684 support
(jcrist/msgspec#563 and more specifically jcrist/msgspec#1026).

Why the dedicated-thread machinery? PEP 734's public
`concurrent.interpreters` only exposes `'isolated'` config, which
refuses to import any C extension missing the
`Py_mod_multiple_interpreters` slot — and that's currently `msgspec`,
which tractor uses pervasively in the IPC layer. So we drop to the
**private** `_interpreters` C module in `'legacy'` config: shared
main GIL, but the `sys.modules`/`__main__`/globals isolation we
actually want for the actor model. We also feature-gate to **3.14+**
via the public-module presence check: the private API exists on 3.13,
but tractor's multi-trio-task usage of `_interpreters.exec()` wedges
silently there and only works cleanly on 3.14.

This PR also documents two known hang classes specific to in-thread
subints (a class-A SIGINT-starvation hang and an orphaned-channel
park) and ships the `skipon_spawn_backend()` pytest marker +
`tractor.devx._debug_hangs` module that fell out of triaging them. It
serves as the foundation for the stacked subint-family PRs (#447
forkserver, future fork variants).

---

### Src of research

- PEP 734 — `concurrent.interpreters` public API
  * https://peps.python.org/pep-0734/
- PEP 684 — per-interpreter GIL / `Py_mod_multiple_interpreters`
  * https://peps.python.org/pep-0684/
- CPython public wrapper (`Lib/concurrent/interpreters/`) *
  https://github.com/python/cpython/tree/main/Lib/concurrent/interpreters
- CPython private C ext (`Modules/_interpretersmodule.c`) *
  https://github.com/python/cpython/blob/main/Modules/_interpretersmodule.c
- `msgspec` PEP 684 upstream tracker — jcrist/msgspec#563
- anyio's `to_interpreter._Worker` for a `set___main___attrs()`
reference impl *
https://github.com/agronholm/anyio/blob/master/src/anyio/to_interpreter.py

---

### Summary of changes

By chronological commit (all hashes from
`main..subint_spawner_backend`),

- ([d318f1f8][d318f1f8]) Add the `'subint'` spawn-method scaffold to
  `tractor.spawn._spawn`: register the key in `SpawnMethodKey`, gate
  `try_set_start_method('subint')` on the public
  `concurrent.interpreters` module, and wire the dispatch table.
- ([b8f243e9][b8f243e9]) Implement min-viable `subint_proc()` in
  `tractor.spawn._subint`: `_interpreters.create('legacy')` →
  bootstrap-string driven `_interpreters.exec()` on a daemon
  `threading.Thread` → reuse `ipc_server.wait_for_peer()` for the
  existing UDS/TCP handshake → graceful `Portal.cancel_actor()`
  teardown → `_interpreters.destroy()`.
- ([8a8d01e0][8a8d01e0]) Doc the `_interpreters` private-API choice
  in the `_subint` module docstring (PEP 734 isolated mode +
  `msgspec` PEP 684 incompat → forced legacy mode + 3.14 floor).
- ([03bf2b93][03bf2b93]) Avoid skipping the `tractor.ipc._ringbuf`
  import path when `cffi` is missing (now an optional `eventfd` dep).
- ([d2ea8aa2][d2ea8aa2]) Handle py3.14+ stdlib incompats as test
  skips so the suite can boot under the new floor.
- ([31cbd11a][31cbd11a]) Fix the subint destroy-race: route
  `_interpreters.exec()` to a **dedicated** `threading.Thread` rather
  than `trio.to_thread.run_sync()`, since trio's thread cache
  recycles workers and leaves the subint tstate attached — blocking
  `_interpreters.destroy()` in teardown.
- ([99541fee][99541fee]) Bound the teardown shields with
  `_HARD_KILL_TIMEOUT=3.0s` + `daemon=True` driver thread +
  `move_on_after()` shielded scopes; abandon-and-log if the subint
  can't be cleanly destroyed (vs. wedging the parent forever).
- ([09466a1e][09466a1e]) Add `tractor.devx._debug_hangs` exposing
  `dump_on_hang()` (file-target
  `faulthandler.dump_traceback_later()`) and
  `track_resource_deltas()` (thread + subint counts) — the
  actually-useful triage primitives that fell out of subint bringup.
- ([34d9d482][34d9d482]) Raise the `subint`-floor to py3.14 in
  `pyproject.toml` and split deps into per-feature groups (`subints`,
  `eventfd`, `sync_pause`) so 3.14-only deps don't bleed into the
  3.13 install path.
- ([2ed5e6a6][2ed5e6a6]) Add `tests/test_subint_cancellation.py` —
  focused cancel + hard-kill audit covering the corner cases the
  hard-kill paths are meant to cover.
- ([4a325458][4a325458]) Doc the two `subint` hang classes
  (`subint_sigint_starvation_issue.md`,
  `subint_cancel_delivery_hang_issue.md`) and arm `dump_on_hang()` in
  the test conftests.
- ([189f4e3f][189f4e3f]) Wall-cap `subint` audit tests via
  `@pytest.mark.timeout(method='thread')` so a regressing hang fails
  fast instead of stalling CI.
- ([a6cbac95][a6cbac95]) Bump `uv.lock` for `pytest-timeout` +
  py3.13-gated wheel deps.
- ([59987745][59987745]) Add a global `200s` `pytest-timeout` ceiling
  in `pyproject.toml`.
- ([985ea76d][985ea76d]) Skip `test_stale_entry_is_deleted` under
  `--spawn-backend=subint` (class-A SIGINT-starvation reproducer;
  tracked in `subint_sigint_starvation_issue.md`).
- ([f3cea714][f3cea714]) Expand the `subint` SIGINT-starvation hang
  catalog with additional reproducers / strace evidence.
- ([3b26b59d][3b26b59d]) Add the
  `@pytest.mark.skipon_spawn_backend('<backend>', reason='...')`
  marker (registered in `tractor._testing.pytest`, expanded in
  `pytest_collection_modifyitems`) so backend-borked tests can be
  cleanly gated.
- ([4b2a0886][4b2a0886]) Apply `skipon_spawn_backend('subint', ...)`
  (with hang-doc tracker refs as `reason=`) to the affected test
  modules.
- ([c041518b][c041518b]) ([a65fded4][a65fded4]) Prompt-IO logs for
  the destroy-race fix and hang-class docs (NLNet generative-AI
  provenance).
- ([9cf3d588][9cf3d588]) ([d3d6f646][d3d6f646]) Reorganize
  `.gitignore` by skill/purpose; ignore notes & snippets subdirs.
- ([ba86d482][ba86d482]) ([b1a0753a][b1a0753a]) `/run-tests` skill
  polish: lastfailed-cache inspection + venv pre-flight expansion.
- ([b524ee46][b524ee46]) ([64ddc42a][64ddc42a]) Bump `xonsh` to
  `0.22.8`-pre and pin to GH `main` editable as a workaround for an
  upstream packaging gap on 3.14.

---

### Scopes changed

- `tractor.spawn._subint` (new, ~433 LoC)
  * `subint_proc()` — full subint actor lifecycle (create /
    driver-thread / handshake / cancel / destroy)
  * `_HARD_KILL_TIMEOUT` + bounded `move_on_after()` shields at every
    teardown blocking-point
  * extensive module docstring covering the PEP 734 / 684 / `msgspec`
    triangle + 3.14 gate rationale

- `tractor.spawn._spawn`
  * register `'subint'` in `SpawnMethodKey` + `_methods` dispatch
  * runtime gate `try_set_start_method('subint')` on the public
    `concurrent.interpreters` module presence

- `tractor._child`
  * factor `_actor_child_main()` shared entry shape used by both the
    CLI (trio/mp subproc backends) and the subint bootstrap string

- `tractor._testing.pytest`
  * `skipon_spawn_backend(*backends, reason=None)` marker +
    `pytest_collection_modifyitems()` expansion
  * convert `try_set_start_method` `RuntimeError` into a clean
    `pytest.UsageError`

- `tractor.devx._debug_hangs` (new)
  * `dump_on_hang()` — file-target
    `faulthandler.dump_traceback_later()` (avoids `pytest`'s stderr
    capture)
  * `track_resource_deltas()` — log thread + subint counts across a
    block

- `pyproject.toml` + `uv.lock`
  * raise `requires-python` floor to `>=3.13, <3.15`
  * split optional-deps into `subints` (msgspec≥0.21), `eventfd`
    (cffi), `sync_pause` (greenback) groups
  * `[tool.uv.dependency-groups]` → `subints = {requires-python =
    ">=3.14"}`
  * add global `pytest-timeout=200s` ceiling

- `ai/conc-anal/` (new)
  * `subint_sigint_starvation_issue.md` — class-A GIL-contention
    SIGINT-never-delivered hang
  * `subint_cancel_delivery_hang_issue.md` — orphaned-channel park
    after subint teardown

- `tests/test_subint_cancellation.py` (new)
  * focused subint cancel + hard-kill test audit

- `tests.discovery.test_registrar`, `tests.test_cancellation`,
`tests.test_inter_peer_cancellation`, `tests.test_pubsub`,
`tests.test_ringbuf`, `tests.test_shm`
  * `pytestmark = pytest.mark.skipon_spawn_backend('subint',
    reason='...')` with hang-doc tracker refs

- `.claude/skills/run-tests/SKILL.md`
  * `lastfailed`-cache inspection + venv pre-flight coverage
    (incidental)

---

### Future follow up

- [ ] **Drop the dedicated-thread machinery once `msgspec` PEP 684 +
  isolated mode lands** Once jcrist/msgspec#563 ships and our
  remaining C-ext deps grow `Py_mod_multiple_interpreters`, switch
  from the private `_interpreters` legacy-mode driver to public
  `concurrent.interpreters.create()` (isolated) +
  `trio.to_thread.run_sync(Interpreter.exec, ...)`. Tracked in #450.

- [ ] **Surface subint-bootstrap exceptions to the parent task via a
  `nonlocal err` slot** `tractor.spawn._subint:_subint_target()`
  currently only `log.exception()`s a hard bootstrap failure (e.g.
  `ImportError` of the actor module inside the subint, or syntax
  error in the bootstrap string). Adopt anyio's `(retval,
  is_exception)` tuple pattern (and the equivalent already used in
  the `_subint_forkserver.py:480-494` block on the stacked PR #447)
  to re-raise into the parent — needs to coordinate with the existing
  `trio.Cancelled` paths around `subint_exited.wait()`. (`?TODO` in
  `_subint.py`.)

- [ ] **Switch bootstrap arg-passing to
  `_interpreters.set___main___attrs()`** The current
  `repr()`-into-bootstrap-string approach only roundtrips literals.
  For a pre-built `SpawnSpec` struct, credentials, or callables we'd
  want anyio's `set___main___attrs(interp_id, {...})` pattern.
  Tracked in #379. (`?TODO` in `_subint.py`.)

- [ ] **Resolve the class-A SIGINT-starvation hang catalog**
  `ai/conc-anal/subint_sigint_starvation_issue.md` documents
  reproducers where an abandoned-subint thread starves the parent's
  trio loop hard enough that **SIGINT is dropped** at the
  kernel↔Python boundary. Currently mitigated via
  `skipon_spawn_backend('subint')` on affected tests + the
  `_HARD_KILL_TIMEOUT` shields. Real fix is upstream-CPython-shaped
  (per-interp GIL / true subint cancellation primitive).

- [ ] **Fix the orphaned-channel park after subint teardown**
  `ai/conc-anal/subint_cancel_delivery_hang_issue.md` — Ctrl-C-able
  (so not the GIL-hostage class) but the parent-side
  `process_messages` loop parks on a now-dead subint's IPC channel
  without ever seeing a clean EOF/`BrokenResourceError`.
  Tractor-shaped fix; needs an explicit channel-tear from the subint
  side as part of the soft-kill sequence.

- [ ] **Cross-link the cancel-cascade hang from PR #447**
  `test_nested_multierrors` cancel-cascade hang gated by pytest
  `--capture=fd` is mostly a `subint_forkserver` symptom (#449) but
  the underlying mechanism may also affect single-process subints
  under specific schedules — worth re-running the audit once #447
  lands.

---

---

(this pr content was generated in some part by [`claude-code`][claude-code-gh])

[d318f1f8]: https://github.com/goodboy/tractor/commit/d318f1f8f4f6a056a8ca67d7f15e4e2a72507ea3
[b8f243e9]: https://github.com/goodboy/tractor/commit/b8f243e98d38688131d56397a87e3e56106bf4a9
[8a8d01e0]: https://github.com/goodboy/tractor/commit/8a8d01e07606086eaddc2ae3a95247567fb3b499
[03bf2b93]: https://github.com/goodboy/tractor/commit/03bf2b931ecea9528e366beefa813b2a3f196781
[d2ea8aa2]: https://github.com/goodboy/tractor/commit/d2ea8aa2deb681faf7986759ff822579ab9ee22f
[31cbd11a]: https://github.com/goodboy/tractor/commit/31cbd11a5b9f309102eac9f7a79157ea6cf0d3bd
[99541fee]: https://github.com/goodboy/tractor/commit/99541feec755d961138995ed54f62fadfa06b0f7
[09466a1e]: https://github.com/goodboy/tractor/commit/09466a1e9dc548040df0bddb5aed96402c1d0a19
[34d9d482]: https://github.com/goodboy/tractor/commit/34d9d482e407f61362e7d5b186a5d7ae95bd426a
[2ed5e6a6]: https://github.com/goodboy/tractor/commit/2ed5e6a6e87069352a300e1ce9cf660f0600fd49
[4a325458]: https://github.com/goodboy/tractor/commit/4a3254583b1168382314fdad1b268c2498bbf729
[189f4e3f]: https://github.com/goodboy/tractor/commit/189f4e3ffc1aad8203b91a59522f923ab3ec3fe4
[a6cbac95]: https://github.com/goodboy/tractor/commit/a6cbac954d60429b9f39809adddd6152bda7f6a9
[59987745]: https://github.com/goodboy/tractor/commit/59987745359e238d78c8de68bc9c502236b4e395
[985ea76d]: https://github.com/goodboy/tractor/commit/985ea76de593d5707f5fcdd7f3095dfca6a77176
[f3cea714]: https://github.com/goodboy/tractor/commit/f3cea714bcca0fc49b8170dab8014ca11008a968
[3b26b59d]: https://github.com/goodboy/tractor/commit/3b26b59dad8c6b365890746ed5acc0811bec94c6
[4b2a0886]: https://github.com/goodboy/tractor/commit/4b2a0886c3ac4b8fdeb1db3db820df2396f3c198
[c041518b]: https://github.com/goodboy/tractor/commit/c041518bdb4aa25fb182ace4bb8cb8c7d49379ab
[a65fded4]: https://github.com/goodboy/tractor/commit/a65fded4c6b545c456736d1a4128c43804e87bf7
[9cf3d588]: https://github.com/goodboy/tractor/commit/9cf3d588e7e25578e82ab39034bfce5b8c21fa1f
[d3d6f646]: https://github.com/goodboy/tractor/commit/d3d6f646f9c48bad8e4e0ca1f09493c1750bc5e3
[ba86d482]: https://github.com/goodboy/tractor/commit/ba86d482e3848ff194e45246b9ac00c69089ccee
[b1a0753a]: https://github.com/goodboy/tractor/commit/b1a0753a3fb0dad536fb33c103f14356e63b70fe
[b524ee46]: https://github.com/goodboy/tractor/commit/b524ee463389982cf0c69a926f25725af6c5004c
[64ddc42a]: https://github.com/goodboy/tractor/commit/64ddc42ad84e90fc47f909a3a4eb49f8cc125842
[claude-code-gh]: https://github.com/anthropics/claude-code

### CODE REVIEW

File: tractor/spawn/_subint.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+# mode) and pick up per-interp-GIL parallelism for free.
+#
+# References:
+# - PEP 734 (`concurrent.interpreters` public API):
+#   https://peps.python.org/pep-0734/
+# - PEP 684 (per-interpreter GIL / `Py_mod_multiple_interpreters`):
+#   https://peps.python.org/pep-0684/
+# - stdlib docs (3.14+):
+#   https://docs.python.org/3.14/library/concurrent.interpreters.html
+# - CPython public wrapper source (`Lib/concurrent/interpreters/`):
```
Comment (goodboy) [inline tractor/spawn/_subint.py:92]: we're missing a link to the older [pep-554](https://peps.python.org/pep-0554/#the-interpreters-module) here?

File: tractor/spawn/_subint.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+from typing import (
+    Any,
+    TYPE_CHECKING,
+)
+
+import trio
+from trio import TaskStatus
+
+
+# NOTE: we reach into the *private* `_interpreters` C module
```
Comment (goodboy) [inline tractor/spawn/_subint.py:53]: Reason for pinning to the legacy API..

File: ai/conc-anal/subint_sigint_starvation_issue.md [hunk 3]
```diff
@@ -0,0 +1,350 @@
+# `subint` backend: abandoned-subint thread can wedge main trio event loop (Ctrl-C unresponsive)
```
Comment (goodboy) [inline ai/conc-anal/subint_sigint_starvation_issue.md:1]: The full reason we need a *GIL per subint*..
